neuralpredictors/layers/readouts/multi_readout.py

```diff
@@ -1,107 +1,147 @@
+
+    Args:
+        in_shape_dict (dict): dictionary of data_key and the corresponding dataset's shape as an output of the core.
+        n_neurons_dict (dict): dictionary of data_key and the corresponding dataset's number of neurons
+        base_readout (torch.nn.Module): base readout class. If None, self._base_readout must be set manually in the inheriting class's definition
+        mean_activity_dict (dict): dictionary of data_key and the corresponding dataset's mean responses. Used to initialize the readout bias with.
+                                   If None, the bias is initialized with 0.
+        clone_readout (bool): whether to clone the first data_key's readout to all other readouts, only allowing for a scale and offset.
+                              This is a rather simple method to enforce parameter-sharing between readouts. For more sophisticated methods,
+                              refer to MultiReadoutSharedParametersBase
+        gamma_readout (float): regularization strength
+        **kwargs:
+    """
+
     _base_readout = None
 
-    def __init__(self, in_shape, loaders, gamma_readout, clone_readout=False, **kwargs):
+    def __init__(
+        self,
+        in_shape_dict,
+        n_neurons_dict,
+        base_readout=None,
+        mean_activity_dict=None,
+        clone_readout=False,
+        gamma_readout=1.0,
+        **kwargs
+    ):
+
+        # The `base_readout` can be overridden only if the static property `_base_readout` is not set
         if self._base_readout is None:
-            raise ValueError("Attribute _base_readout must be set")
+            self._base_readout = base_readout
 
+        if self._base_readout is None:
+            raise ValueError("Attribute _base_readout must be set")
         super().__init__()
 
-        self.in_shape = in_shape
-        self.neurons = OrderedDict([(k, loader.dataset.n_neurons) for k, loader in loaders.items()])
-        if "positive" in kwargs:
-            self._positive = kwargs["positive"]
+        for i, data_key in enumerate(n_neurons_dict):
+            first_data_key = data_key if i == 0 else first_data_key
+            mean_activity = mean_activity_dict[data_key] if mean_activity_dict is not None else None
 
-        self.gamma_readout = gamma_readout  # regularisation strength
+            readout_kwargs = self.prepare_readout_kwargs(i, data_key, first_data_key, **kwargs)
 
-        for i, (k, n_neurons) in enumerate(self.neurons.items()):
             if i == 0 or clone_readout is False:
                 self.add_module(
-                    k,
-                    self._base_readout(in_shape=in_shape, outdims=n_neurons, **kwargs),
+                    data_key,
+                    self._base_readout(
+                        in_shape=in_shape_dict[data_key],
+                        outdims=n_neurons_dict[data_key],
+                        mean_activity=mean_activity,
+                        reg_weight=gamma_readout,
```

Comment:
Based on the comment above this would be removed.

Reply:
you mean line 62?

Reply:
yes exactly

Reply:
I also have another question here. When initializing each readout module, you pass in `in_shape` and `outdims` followed by `mean_activity`. But then, for instance, in [FullGaussian2d](https://github.com/sinzlab/neuralpredictors/blob/4ac2ed8d6404eef52bdf619ed1a9fe565994f497/neuralpredictors/layers/readouts/gaussian.py#L205), the first three arguments are positional (i.e. must be passed) and one of them is `bias`. I know you said you tested the initialization and it works, but I am just wondering how? I thought initializing a FullGaussian2d should throw an error because `bias`  was not passed as input.

Reply:
All arguments are passed as keyword arguments (`inshape`, `outdims`, `mean_activity`, `reg_weight`) and the `**readout_kwargs` results in keyword arguments as well. So if everything that is passed is a keyword argument, the order doesn't matter. If you choose a `base_readout` that needs a specific argument (like `bias` in `FullGaussian2d`) and you don't give it in the `readout_kwargs`, then of course it will give you an error